packages/nimble-components/src/table/models/keyboard-navigation-manager.ts

```diff
@@ -0,0 +1,1243 @@
+        rowElements: TableRowFocusableElements,
+        cellContentIndex: number,
+        columnIndex?: number,
+        rowIndex?: number
+    ): boolean {
+        const newColumnIndex = columnIndex ?? this.columnIndex;
+        const newRowIndex = rowIndex ?? this.rowIndex;
+
+        if (
+            newColumnIndex >= 0
+            && newColumnIndex < rowElements.cells.length
+            && cellContentIndex >= 0
+            && cellContentIndex
+                < rowElements.cells[newColumnIndex]!.cell.cellView
+                    .tabbableChildren.length
+        ) {
+            this.setCellContentFocusState(
+                cellContentIndex,
+                newRowIndex,
+                newColumnIndex,
+                true
+            );
+            return true;
+        }
+
+        return false;
+    }
+
+    private trySetCellActionMenuFocus(
+        rowElements: TableRowFocusableElements,
+        columnIndex?: number,
+        rowIndex?: number
+    ): boolean {
+        const newColumnIndex = columnIndex ?? this.columnIndex;
+        const newRowIndex = rowIndex ?? this.rowIndex;
+
+        if (
+            newColumnIndex >= 0
+            && newColumnIndex < rowElements.cells.length
+            && rowElements.cells[newColumnIndex]!.actionMenuButton
+        ) {
+            this.setCellActionMenuFocusState(newRowIndex, newColumnIndex, true);
+            return true;
+        }
+
+        return false;
+    }
+
+    private setCellActionMenuFocusState(
+        rowIndex: number,
+        columnIndex: number,
+        focusElement: boolean
+    ): void {
+        this.focusType = TableFocusType.cellActionMenu;
+        this.setRowCellFocusState(columnIndex, rowIndex, focusElement);
+    }
+
+    private setCellContentFocusState(
+        cellContentIndex: number,
+        rowIndex: number,
+        columnIndex: number,
+        focusElement: boolean
+    ): void {
+        this.focusType = TableFocusType.cellContent;
+        this.cellContentIndex = cellContentIndex;
+        this.setRowCellFocusState(columnIndex, rowIndex, focusElement);
+    }
+
+    private setRowFocusState(rowIndex?: number): void {
+        this.focusType = TableFocusType.row;
+        if (rowIndex !== undefined) {
+            this.rowIndex = rowIndex;
+        }
+        this.updateNavigationMode();
+    }
+
+    private setCellFocusState(
+        columnIndex: number,
+        rowIndex: number,
+        focusElement: boolean
+    ): void {
+        this.focusType = TableFocusType.cell;
+        this.setRowCellFocusState(columnIndex, rowIndex, focusElement);
+    }
+
+    private setRowCellFocusState(
+        columnIndex: number,
+        rowIndex: number,
+        focusElement: boolean
+    ): void {
+        this.rowIndex = rowIndex;
+        this.columnIndex = columnIndex;
+        this.updateNavigationMode();
+        if (focusElement) {
+            this.focusCurrentRow(true);
+        }
+    }
+
+    private updateNavigationMode(): void {
```

Comment:
If all this method does is change a mode purely based on the current state of another member, would it be better to just make `focusType` notify when it changes, and then just update `inNavigationMode` there? That way, one doesn't have to know when to just call this method. There are certain call-sites for this method where it's unclear why it is needed (namely when `focusType` doesn't seem to be being set in its immediate vicinity).

Reply:
Since `inNavigationMode` directly maps to certain `focusType` values, I just replaced it with a derived property getter, and got rid of this method.